miden/src/repl/mod.rs

```diff
@@ -0,0 +1,317 @@
+        );
+    }
+
+    println!("========================== Miden REPL ============================");
+    println!();
+    println!("Available commands:");
+    println!();
+    println!("!stack: displays the complete state of the stack");
+    println!("!mem: displays the state of the entire memory");
+    println!("!mem[i]: displays the state of the memory at address i");
+    println!("!undo: remove the last instruction");
+    println!("!program: display the program");
+    println!("!help: prints out all the available commands");
+
+    let mut stack_flag = false;
+
+    // state of the entire memory at the latest clk cycle.
+    let mut memory: Vec<(u64, Word)> = Vec::new();
+
+    // initialising readline.
+    let mut rl = Editor::<()>::new().unwrap();
+    loop {
+        let program = format!(
+            "begin\n{}\nend",
+            program_lines
+                .iter()
+                .map(|l| format!("    {}", l))
+                .collect::<Vec<_>>()
+                .join("\n")
+        );
+        let program_with_procs = format!("{}\n{}", functions_miden, program.clone());
+
+        let result = execute(program_with_procs, vec![]);
+
+        let mut result_string = "0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0".to_string();
+        if !program_lines.is_empty() {
+            match result {
+                Ok((mem, stack_state)) => {
+                    result_string = stack_state;
+                    if stack_flag {
+                        println!("{}", result_string);
+                    }
+                    memory = mem;
+                }
+                Err(e) => {
+                    println!("{}", format!("Error running program: {:?}", e));
+                    program_lines.pop();
+                }
+            }
+        }
+        let readline = rl.readline(">> ");
+        match readline {
+            Ok(line) => {
+                if line == "!program" {
+                    println!("{}", program);
+                    stack_flag = false;
+                } else if line == "!help" {
+                    println!("Available commands:");
+                    println!();
+                    println!("!stack: displays the complete state of the stack");
+                    println!("!mem: displays the state of the entire memory");
+                    println!("!mem[i]: displays the state of the memory at address i");
+                    println!("!undo: remove the last instruction");
+                    println!("!program: display the program");
+                    println!("!help: prints out all the available commands");
+                    stack_flag = false;
+                } else if line == "!mem" {
+                    for (addr, mem) in &memory {
+                        // each memory register converted from Felt to an integer.
+                        let mem_int = mem.iter().map(|&x| x.as_int()).collect::<Vec<_>>();
+                        println!("{}: {:?}", addr, mem_int)
```

Comment:
This code is used a couple of time. I'd put it into a separate helper function - e.g., `print_mem_address()`.